File: clients/ts/src/ledger.ts
import { closeSync, existsSync, fsyncSync, mkdirSync, openSync, readFileSync, renameSync, writeSync } from "node:fs";
import { dirname } from "node:path";

// Prepaid balances a seller holds for its buyers. The file format is the
// Go paywall's (paywall.FileLedger), so a seller can move between them.

export interface Withdrawal {
  id: string;
  account: string;
  requested: string; // "all" or uaeth, as asked
  amount: string; // uaeth
  status: "reserved" | "pending" | "confirmed";
  txHash?: string;
  /** The signed payout (base64), kept from before it's broadcast until it's in a block. */
  txBytes?: string;
  sequence?: number;
  at: string;
  /** When the payout's sequence was first seen used while the payout wasn't on chain. */
  sequenceSpentAt?: string;
}

/** One transfer of what an aether-pull buyer owes (the Go paywall's Collection). */
export interface Collection {
  id: string;
  account: string;
  amount: string; // uaeth
  status: "reserved" | "pending" | "confirmed" | "failed";
  txHash?: string;
  /** The signed transfer (base64), kept from before it's broadcast until it's in a block. */
  txBytes?: string;
  sequence?: number;
  at: string;
  sequenceSpentAt?: string;
  log?: string;
}

/** Where one aether-pull buyer stands, in uaeth. */
export interface PullAccount {
  accrued: bigint; // charged, not yet being collected
  inFlight: bigint; // in the open collection
  unpaid: bigint; // from failed collections
}

export const owedOf = (a: PullAccount) => a.accrued + a.inFlight + a.unpaid;

/** What aether-pull buyers owe. Every method must be atomic and durable. */
export interface PullLedger {
  /** Charges amount for a requestId not charged before; ok false (nothing changes) if what's accrued would exceed limit. */
  accrue(account: string, requestId: string, amount: bigint, limit: bigint, forgetAfterMs: number): { accrued: bigint; ok: boolean; fresh: boolean };
  /** Reverses an accrue not yet taken into a collection. */
  unaccrue(account: string, requestId: string, amount: bigint): void;
  pullAccount(account: string): PullAccount;
  /** Accounts with something accrued or a collection open. */
  collectable(): string[];
  /** The account's open collection, or a new one for everything accrued (undefined: nothing to collect). */
  openCollection(account: string, id: string, atMs: number): Collection | undefined;
  saveCollection(c: Collection): void;
  /** Ends the open collection: collected, or failed (its amount becomes unpaid). */
  closeCollection(account: string, collected: boolean, log?: string): void;
  /** Moves unpaid back to accrued, to collect again. */
  reinstate(account: string): void;
}

export class LedgerError extends Error {
  constructor(readonly code: "insufficient" | "below_minimum", message: string) {
    super(message);
  }
}

/** Every method must be atomic and durable: these are customers' funds. */
export interface Ledger {
  /** Adds a deposit to account, once per transaction hash. */
  credit(depositTx: string, account: string, amount: bigint): { balance: bigint; credited: boolean };
  /** Deducts amount for a requestId not charged before; ok false (nothing changes) if the balance is short. */
  charge(account: string, requestId: string, amount: bigint, forgetAfterMs: number): { balance: bigint; ok: boolean; fresh: boolean };
  /** Reverses a charge, forgetting the requestId. */
  refund(account: string, requestId: string, amount: bigint): void;
  balance(account: string): bigint;
  /** Deducts a withdrawal (undefined amount: the whole balance) and records it under id, once. Throws LedgerError. */
  reserveWithdrawal(account: string, id: string, amount: bigint | undefined, min: bigint, atMs: number): { withdrawal: Withdrawal; balance: bigint; fresh: boolean };
  saveWithdrawal(w: Withdrawal): void;
  /** Returns a reserved withdrawal to the balance and forgets it: only for a payout that can never land. */
  cancelWithdrawal(account: string, id: string): void;
}

interface State {
  balances: Record<string, string>;
  deposits: Record<string, { account: string; amount: string; at: string }>;
  requests: Record<string, { amount: string; forgetAfter: string }>;
  withdrawals?: Record<string, Withdrawal>;
  pull?: Record<string, { accrued?: string; unpaid?: string; open?: Collection }>;
  pullRequests?: Record<string, { amount: string; forgetAfter: string }>;
  collections?: Record<string, Collection>;
}

const UINT = /^[0-9]+$/;
const num = (s: string | undefined) => (s && UINT.test(s) ? BigInt(s) : 0n);

/**
 * A Ledger in one JSON file, rewritten atomically (temp file, fsync,
 * rename) on every change. For one server process; several need a shared
 * Ledger (e.g. a database). path "" keeps it in memory (tests).
 */
export class FileLedger implements Ledger, PullLedger {
  private state: State = { balances: {}, deposits: {}, requests: {}, withdrawals: {} };

  constructor(private readonly path: string) {
    if (path && existsSync(path)) {
      try {
        this.state = JSON.parse(readFileSync(path, "utf8")) as State;
      } catch (e) {
        throw new Error(`prepaid ledger ${path} is unreadable: ${(e as Error).message}`);
      }
      this.state.balances ??= {};
      this.state.deposits ??= {};
      this.state.requests ??= {};
      this.state.withdrawals ??= {};
    }
    this.state.pull ??= {};
    this.state.pullRequests ??= {};
    this.state.collections ??= {};
  }

  // --- aether-pull ---

  pullAccount(account: string): PullAccount {
    const r = this.state.pull![account];
    return { accrued: num(r?.accrued), inFlight: num(r?.open?.amount), unpaid: num(r?.unpaid) };
  }

  private rec(account: string) {
    return (this.state.pull![account] ??= {});
  }

  private dropIfEmpty(account: string) {
    const r = this.state.pull![account];
    if (r && !r.accrued && !r.unpaid && !r.open) delete this.state.pull![account];
  }

  accrue(account: string, requestId: string, amount: bigint, limit: bigint, forgetAfterMs: number) {
    const key = `${account}/${requestId}`;
    const accrued = this.pullAccount(account).accrued;
    if (this.state.pullRequests![key]) return { accrued, ok: false, fresh: false };
    if (accrued + amount > limit) return { accrued, ok: false, fresh: true };
    return this.mutate(() => {
      this.rec(account).accrued = (accrued + amount).toString();
      this.state.pullRequests![key] = { amount: amount.toString(), forgetAfter: new Date(forgetAfterMs).toISOString() };
      return { accrued: accrued + amount, ok: true, fresh: true };
    });
  }

  unaccrue(account: string, requestId: string, amount: bigint) {
    const key = `${account}/${requestId}`;
    if (!this.state.pullRequests![key]) throw new Error(`request ${key} was not charged`);
    const accrued = this.pullAccount(account).accrued;
    if (accrued < amount) throw new Error(`request ${key} is already being collected`);
    this.mutate(() => {
      const r = this.rec(account);
      r.accrued = accrued - amount ? (accrued - amount).toString() : undefined;
      if (!r.accrued) delete r.accrued;
      delete this.state.pullRequests![key];
      this.dropIfEmpty(account);
    });
  }

  collectable(): string[] {
    return Object.entries(this.state.pull!).filter(([, r]) => r.accrued || r.open).map(([a]) => a).sort();
  }

  openCollection(account: string, id: string, atMs: number): Collection | undefined {
    const r = this.state.pull![account];
    if (!r) return undefined;
    if (r.open) return { ...r.open };
    if (!r.accrued) return undefined;
    return this.mutate(() => {
      const c: Collection = { id, account, amount: r.accrued!, status: "reserved", at: new Date(atMs).toISOString() };
      r.open = c;
      delete r.accrued;
      return { ...c };
    });
  }

  saveCollection(c: Collection) {
    const r = this.state.pull![c.account];
    if (!r?.open || r.open.id !== c.id) throw new Error(`collection ${c.id} is not open`);
    this.mutate(() => {
      r.open = { ...c };
    });
  }

  closeCollection(account: string, collected: boolean, log?: string) {
    const r = this.state.pull![account];
    if (!r?.open) throw new Error(`${account} has no open collection`);
    this.mutate(() => {
      const c: Collection = { ...r.open!, txBytes: undefined, status: collected ? "confirmed" : "failed", ...(log ? { log } : {}) };
      if (!collected) r.unpaid = (num(r.unpaid) + num(c.amount)).toString();
      this.state.collections![c.id] = c;
      delete r.open;
      this.dropIfEmpty(account);
    });
  }

  reinstate(account: string) {
    const r = this.state.pull![account];
    if (!r?.unpaid) return;
    this.mutate(() => {
      r.accrued = (num(r.accrued) + num(r.unpaid)).toString();
      delete r.unpaid;
    });
  }

  balance(account: string): bigint {
    return num(this.state.balances[account]);
  }

  credit(depositTx: string, account: string, amount: bigint) {
    if (this.state.deposits[depositTx]) return { balance: this.balance(account), credited: false };
    return this.mutate(() => {
      const balance = this.balance(account) + amount;
      this.state.balances[account] = balance.toString();
      this.state.deposits[depositTx] = { account, amount: amount.toString(), at: new Date().toISOString() };
      return { balance, credited: true };
    });
  }

  charge(account: string, requestId: string, amount: bigint, forgetAfterMs: number) {
    const key = `${account}/${requestId}`;
    const bal = this.balance(account);
    if (this.state.requests[key]) return { balance: bal, ok: false, fresh: false };
    if (bal < amount) return { balance: bal, ok: false, fresh: true };
    return this.mutate(() => {
      this.state.balances[account] = (bal - amount).toString();
      this.state.requests[key] = { amount: amount.toString(), forgetAfter: new Date(forgetAfterMs).toISOString() };
      return { balance: bal - amount, ok: true, fresh: true };
    });
  }

  refund(account: string, requestId: string, amount: bigint) {
    const key = `${account}/${requestId}`;
    if (!this.state.requests[key]) return;
    this.mutate(() => {
      this.state.balances[account] = (this.balance(account) + amount).toString();
      delete this.state.requests[key];
    });
  }

  reserveWithdrawal(account: string, id: string, amount: bigint | undefined, min: bigint, atMs: number) {
    const key = `${account}/${id}`;
    const bal = this.balance(account);
    const seen = this.state.withdrawals![key];
    if (seen) return { withdrawal: seen, balance: bal, fresh: false };
    const take = amount ?? bal;
    if (bal <= 0n || take <= 0n || take > bal) throw new LedgerError("insufficient", "insufficient balance");
    if (take < min && take !== bal) throw new LedgerError("below_minimum", "below the minimum withdrawal");
    return this.mutate(() => {
      const withdrawal: Withdrawal = {
        id, account, requested: amount === undefined ? "all" : amount.toString(), amount: take.toString(), status: "reserved", at: new Date(atMs).toISOString(),
      };
      this.state.balances[account] = (bal - take).toString();
      this.state.withdrawals![key] = withdrawal;
      return { withdrawal, balance: bal - take, fresh: true };
    });
  }

  saveWithdrawal(w: Withdrawal) {
    const key = `${w.account}/${w.id}`;
    if (!this.state.withdrawals![key]) throw new Error("withdrawal not found");
    this.mutate(() => {
      this.state.withdrawals![key] = { ...w };
    });
  }

  cancelWithdrawal(account: string, id: string) {
    const key = `${account}/${id}`;
    const w = this.state.withdrawals![key];
    if (!w) throw new Error("withdrawal not found");
    this.mutate(() => {
      this.state.balances[account] = (this.balance(account) + num(w.amount)).toString();
      delete this.state.withdrawals![key];
    });
  }

  /** Applies fn and saves; on a failed save, the change is undone. */
  private mutate<T>(fn: () => T): T {
    const snapshot = structuredClone(this.state);
    try {
      const r = fn();
      this.save();
      return r;
    } catch (e) {
      this.state = snapshot;
      throw e;
    }
  }

  private save() {
    const now = Date.now();
    for (const m of [this.state.requests, this.state.pullRequests!]) {
      for (const [k, r] of Object.entries(m)) {
        if (!r.forgetAfter.startsWith("0001-") && Date.parse(r.forgetAfter) < now) delete m[k];
      }
    }
    if (!this.path) return;
    mkdirSync(dirname(this.path), { recursive: true, mode: 0o700 });
    const tmp = this.path + ".tmp";
    const fd = openSync(tmp, "w", 0o600);
    try {
      // Like the Go ledger: empty aether-pull maps are left out.
      const out: Partial<State> = { ...this.state };
      for (const k of ["pull", "pullRequests", "collections"] as const) if (!Object.keys(out[k] ?? {}).length) delete out[k];
      writeSync(fd, JSON.stringify(out, null, 2));
      fsyncSync(fd);
    } finally {
      closeSync(fd);
    }
    renameSync(tmp, this.path);
  }
}

File: clients/python/aether_client/ledger.py
"""Prepaid balances a seller holds for its buyers.

The file format is the Go paywall's (paywall.FileLedger), so a seller can
move between them.
"""

import copy
import json
import os
import re
import threading
import time
from datetime import datetime, timezone
from typing import Optional

_UINT = re.compile(r"[0-9]+")


def _num(s) -> int:
    return int(s) if isinstance(s, str) and _UINT.fullmatch(s) else 0


def _iso(ts: float) -> str:
    return datetime.fromtimestamp(ts, timezone.utc).strftime("%Y-%m-%dT%H:%M:%S.%fZ")


def _parse_time(s: str) -> Optional[float]:
    if not s or s.startswith("0001-"):
        return None
    m = re.fullmatch(r"(\d{4}-\d\d-\d\dT\d\d:\d\d:\d\d)(?:\.(\d+))?(Z|[+-]\d\d:\d\d)", s)
    if not m:
        return None
    base = datetime.strptime(m.group(1), "%Y-%m-%dT%H:%M:%S")
    tz = timezone.utc if m.group(3) == "Z" else datetime.strptime(m.group(3).replace(":", ""), "%z").tzinfo
    frac = float("0." + m.group(2)) if m.group(2) else 0.0
    return base.replace(tzinfo=tz).timestamp() + frac


class PullAccount:
    """Where one aether-pull buyer stands, in uaeth."""

    def __init__(self, accrued=0, in_flight=0, unpaid=0):
        self.accrued, self.in_flight, self.unpaid = accrued, in_flight, unpaid  # charged; being collected; from failed collections

    @property
    def owed(self) -> int:
        return self.accrued + self.in_flight + self.unpaid


class LedgerError(Exception):
    def __init__(self, code: str, message: str):
        super().__init__(message)
        self.code = code  # insufficient | below_minimum


class FileLedger:
    """A ledger in one JSON file, rewritten atomically (temp file, fsync,
    rename) on every change. Every method is atomic: these are customers'
    funds. For one server process; several need a shared ledger (e.g. a
    database) with the same methods. path "" keeps it in memory (tests)."""

    def __init__(self, path: str):
        self.path = path
        self._lock = threading.RLock()
        self.state = {"balances": {}, "deposits": {}, "requests": {}, "withdrawals": {}}
        if path and os.path.exists(path):
            try:
                with open(path) as f:
                    self.state = json.load(f)
            except ValueError as e:
                raise ValueError(f"prepaid ledger {path} is unreadable: {e}") from None
            for k in ("balances", "deposits", "requests", "withdrawals"):
                if not isinstance(self.state.get(k), dict):
                    self.state[k] = {}
        for k in ("pull", "pullRequests", "collections"):
            if not isinstance(self.state.get(k), dict):
                self.state[k] = {}

    # --- aether-pull: what buyers owe (the Go paywall's PullLedger) ---

    def pull_account(self, account: str) -> PullAccount:
        with self._lock:
            r = self.state["pull"].get(account) or {}
            return PullAccount(_num(r.get("accrued")), _num((r.get("open") or {}).get("amount")), _num(r.get("unpaid")))

    def _drop_if_empty(self, account: str):
        r = self.state["pull"].get(account)
        if r is not None and not r.get("accrued") and not r.get("unpaid") and not r.get("open"):
            del self.state["pull"][account]

    def accrue(self, account: str, request_id: str, amount: int, limit: int, forget_after: float):
        """Charges amount for a request_id not charged before: (accrued, ok, fresh). ok is False
        (nothing changes) if what's accrued would exceed limit."""
        with self._lock:
            key = f"{account}/{request_id}"
            accrued = self.pull_account(account).accrued
            if key in self.state["pullRequests"]:
                return accrued, False, False
            if accrued + amount > limit:
                return accrued, False, True

            def change():
                self.state["pull"].setdefault(account, {})["accrued"] = str(accrued + amount)
                self.state["pullRequests"][key] = {"amount": str(amount), "forgetAfter": _iso(forget_after)}
                return accrued + amount, True, True
            return self._mutate(change)

    def unaccrue(self, account: str, request_id: str, amount: int):
        """Reverses an accrue not yet taken into a collection."""
        with self._lock:
            key = f"{account}/{request_id}"
            if key not in self.state["pullRequests"]:
                raise KeyError(f"request {key} was not charged")
            accrued = self.pull_account(account).accrued
            if accrued < amount:
                raise ValueError(f"request {key} is already being collected")

            def change():
                r = self.state["pull"].setdefault(account, {})
                if accrued - amount:
                    r["accrued"] = str(accrued - amount)
                else:
                    r.pop("accrued", None)
                del self.state["pullRequests"][key]
                self._drop_if_empty(account)
            self._mutate(change)

    def collectable(self):
        """Accounts with something accrued or a collection open."""
        with self._lock:
            return sorted(a for a, r in self.state["pull"].items() if r.get("accrued") or r.get("open"))

    def open_collection(self, account: str, cid: str, at: float) -> Optional[dict]:
        """The account's open collection, or a new one for everything accrued (None: nothing to collect)."""
        with self._lock:
            r = self.state["pull"].get(account)
            if not r:
                return None
            if r.get("open"):
                return dict(r["open"])
            if not r.get("accrued"):
                return None

            def change():
                c = {"id": cid, "account": account, "amount": r["accrued"], "status": "reserved", "at": _iso(at)}
                r["open"] = c
                del r["accrued"]
                return dict(c)
            return self._mutate(change)

    def save_collection(self, c: dict):
        with self._lock:
            r = self.state["pull"].get(c["account"])
            if not r or not r.get("open") or r["open"]["id"] != c["id"]:
                raise KeyError(f"collection {c['id']} is not open")

            def change():
                r["open"] = {k: v for k, v in c.items() if v is not None}
            self._mutate(change)

    def close_collection(self, account: str, collected: bool, log: str = ""):
        """Ends the open collection: collected, or failed (its amount becomes unpaid)."""
        with self._lock:
            r = self.state["pull"].get(account)
            if not r or not r.get("open"):
                raise KeyError(f"{account} has no open collection")

            def change():
                c = {k: v for k, v in r["open"].items() if k != "txBytes"}
                c["status"] = "confirmed" if collected else "failed"
                if log:
                    c["log"] = log
                if not collected:
                    r["unpaid"] = str(_num(r.get("unpaid")) + _num(c["amount"]))
                self.state["collections"][c["id"]] = c
                del r["open"]
                self._drop_if_empty(account)
            self._mutate(change)

    def reinstate(self, account: str):
        """Moves unpaid back to accrued, to collect again."""
        with self._lock:
            r = self.state["pull"].get(account)
            if not r or not r.get("unpaid"):
                return

            def change():
                r["accrued"] = str(_num(r.get("accrued")) + _num(r["unpaid"]))
                del r["unpaid"]
            self._mutate(change)

    def balance(self, account: str) -> int:
        with self._lock:
            return _num(self.state["balances"].get(account))

    def credit(self, deposit_tx: str, account: str, amount: int):
        """Adds a deposit to account, once per transaction hash: (balance, credited)."""
        with self._lock:
            if deposit_tx in self.state["deposits"]:
                return self.balance(account), False

            def change():
                bal = self.balance(account) + amount
                self.state["balances"][account] = str(bal)
                self.state["deposits"][deposit_tx] = {"account": account, "amount": str(amount), "at": _iso(_now())}
                return bal, True
            return self._mutate(change)

    def charge(self, account: str, request_id: str, amount: int, forget_after: float):
        """Deducts amount for a request ID not charged before: (balance, ok, fresh)."""
        with self._lock:
            key = f"{account}/{request_id}"
            bal = self.balance(account)
            if key in self.state["requests"]:
                return bal, False, False
            if bal < amount:
                return bal, False, True

            def change():
                self.state["balances"][account] = str(bal - amount)
                self.state["requests"][key] = {"amount": str(amount), "forgetAfter": _iso(forget_after)}
                return bal - amount, True, True
            return self._mutate(change)

    def refund(self, account: str, request_id: str, amount: int):
        """Reverses a charge, forgetting the request ID."""
        with self._lock:
            key = f"{account}/{request_id}"
            if key not in self.state["requests"]:
                return

            def change():
                self.state["balances"][account] = str(self.balance(account) + amount)
                del self.state["requests"][key]
            self._mutate(change)

    def reserve_withdrawal(self, account: str, wid: str, amount: Optional[int], minimum: int, at: float):
        """Deducts a withdrawal (amount None: the whole balance) and records it
        under wid, once: (withdrawal, balance, fresh). Raises LedgerError."""
        with self._lock:
            key = f"{account}/{wid}"
            bal = self.balance(account)
            seen = self.state["withdrawals"].get(key)
            if seen:
                return dict(seen), bal, False
            take = bal if amount is None else amount
            if bal <= 0 or take <= 0 or take > bal:
                raise LedgerError("insufficient", "insufficient balance")
            if take < minimum and take != bal:
                raise LedgerError("below_minimum", "below the minimum withdrawal")

            def change():
                w = {"id": wid, "account": account, "requested": "all" if amount is None else str(amount),
                     "amount": str(take), "status": "reserved", "at": _iso(at)}
                self.state["balances"][account] = str(bal - take)
                self.state["withdrawals"][key] = w
                return dict(w), bal - take, True
            return self._mutate(change)

    def save_withdrawal(self, w: dict):
        with self._lock:
            key = f"{w['account']}/{w['id']}"
            if key not in self.state["withdrawals"]:
                raise KeyError("withdrawal not found")

            def change():
                self.state["withdrawals"][key] = {k: v for k, v in w.items() if v is not None}
            self._mutate(change)

    def cancel_withdrawal(self, account: str, wid: str):
        """Returns a reserved withdrawal to the balance and forgets it: only for a payout that can never land."""
        with self._lock:
            key = f"{account}/{wid}"
            w = self.state["withdrawals"].get(key)
            if not w:
                raise KeyError("withdrawal not found")

            def change():
                self.state["balances"][account] = str(self.balance(account) + _num(w.get("amount")))
                del self.state["withdrawals"][key]
            self._mutate(change)

    def _mutate(self, fn):
        snapshot = copy.deepcopy(self.state)
        try:
            r = fn()
            self._save()
            return r
        except BaseException:
            self.state = snapshot
            raise

    def _save(self):
        now = _now()
        for m in (self.state["requests"], self.state["pullRequests"]):
            for k, r in list(m.items()):
                t = _parse_time(r.get("forgetAfter", ""))
                if t is not None and t < now:
                    del m[k]
        if not self.path:
            return
        d = os.path.dirname(os.path.abspath(self.path))
        os.makedirs(d, mode=0o700, exist_ok=True)
        tmp = self.path + ".tmp"
        fd = os.open(tmp, os.O_WRONLY | os.O_CREAT | os.O_TRUNC, 0o600)
        with os.fdopen(fd, "wb") as f:
            # Like the Go ledger: empty aether-pull maps are left out.
            out = {k: v for k, v in self.state.items() if v or k not in ("pull", "pullRequests", "collections")}
            f.write(json.dumps(out, indent=2).encode())
            f.flush()
            os.fsync(f.fileno())
        os.replace(tmp, self.path)


def _now() -> float:
    return time.time()
